kale/embed/mpca.py

```diff
@@ -1,49 +1,76 @@
 # =============================================================================
 # Author: Shuo Zhou, szhou20@sheffield.ac.uk
 #         Haiping Lu, h.lu@sheffield.ac.uk
-#
-# Python implement of Multilinear Principal Component Analysis (MPCA)
-#
-# Ref: Haiping Lu, K.N. Plataniotis, and A.N. Venetsanopoulos, "MPCA: Multilinear
-# Principal Component Analysis of Tensor Objects", IEEE Transactions on Neural 
-# Networks, Vol. 19, No. 1, Page: 18-39, January 2008.
 # =============================================================================
+
+"""Python implementation of Multilinear Principal Component Analysis (MPCA)
+Includeing implementaion as a scikit-learn object and an independent function.
+
+Reference:
+    Haiping Lu, K.N. Plataniotis, and A.N. Venetsanopoulos, "MPCA: Multilinear
+    Principal Component Analysis of Tensor Objects", IEEE Transactions on Neural 
+    Networks, Vol. 19, No. 1, Page: 18-39, January 2008. For initial Matlab
+    implementation, please go to https://uk.mathworks.com/matlabcentral/fileexchange/26168.
+"""
+
 import numpy as np
-import tensorly as tl
-from tensorly.base import fold, unfold
-from tensorly.tenalg import mode_dot, multi_mode_dot
+# import tensorly as tl
+from tensorly.base import unfold
+from tensorly.tenalg import multi_mode_dot
 from sklearn.base import BaseEstimator, TransformerMixin
 
 
 class MPCA(BaseEstimator, TransformerMixin):
-    def __init__(self, variance_explained=0.97, max_iter=1):
-        """
-        Implementation compatible with sickit-learn
-        ----------
-        Parameter:
-            variance_explained: ration of variance to keep (between 0 and 1)
-            max_iter: max number of iteration, integer
-        ----------
-        Attributes:
-            tPs: list of projection matrices
-        """
-        self.var_exp = variance_explained
+    """MPCA implementation compatible with sickit-learn
+
+    Args:
+        var_explained (float, optional): ration of variance to keep
+            (between 0 and 1). Defaults to 0.97.
+        max_iter (int, optional): max number of iteration. Defaults to 1.
+
+    Attributes:
```

Comment:
It seems that the sorting order (odrIdx in matlab) is not returned. If we fit to 100 but want to test performance w.r.t. 10, 20, 30, how to find the first 10/20/30 Multilinear PCs? With odrIdx, I can just take the first 10/20/30. With the current code, do I have to fit again three times?

Reply:
How about making odrIdx an attribution of MPCA object? Just like tPs in the current implementation.

Reply:
Yes, I think it is good. This also highlights our deep understanding of MPCA. Without ordIdx, MPCA won't be flexible. 

Ranked/ordered features are a good characteristic of PCA/MPCA, e.g. compared to ICA or Autoencoder, in which the features do not have a natural ordering. 

Of course, ordIdx is actually a univariate feature selection process (unsupervised version by default, sort by variance). But PCA has a natural ordering so it will be good for MPCA to have it naturally.